### PR TITLE
Post-pnr static timing anlysis, post-pnr pipelining, and post-pnr visualization

### PR DESCRIPTION
**Pull request by jack-melchert**

### CODE REVIEW

File: archipelago/pnr_.py [hunk 1]
```diff
@@ -116,7 +127,7 @@ def pnr(arch, input_netlist=None, packed_file="", cwd="", app_name="",
         if wave_filename is not None:
             shutil.copy2(wave_filename, copy_to_dir)
 
-    return placement_result, routing_result
+    return placement_result, routing_result, id_to_name
```
Comment (Kuree) [inline archipelago/pnr_.py:130]: Does this mean that the upstream needs to be changed?

File: archipelago/sta.py [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+    print("Loading routing")
+    routing = io.load_routing_result(routing_file)
+
+    app_dir = os.path.dirname(netlist_file)
+
+    sta_analysis(app_dir, placement, routing, id_to_name)
+
+    
+if __name__ == "__main__":
+    sys.path.append(os.path.dirname(os.path.realpath(__file__)))
```
Comment (Kuree) [inline archipelago/sta.py:42]: Why is this needed?

File: setup.py [hunk 3]
```diff
@@ -14,5 +14,6 @@
     install_requires=[
         "pythunder",
         "pycyclone",
+        "pillow",
```
Comment (Kuree) [inline setup.py:17]: This is not a hard requirement. Can you make it optional so people can install it as `archipelago[vis]` or something similar?

File: archipelago/pipeline.py [hunk 4]
```diff
@@ -0,0 +1,1028 @@
+import sys
+import os
+import argparse
+import re
+import itertools
+import glob
+import json
+ 
+from graphviz import Digraph
```
Comment (Kuree) [inline archipelago/pipeline.py:9]: See my comment below about `setup.py` requirements.

File: archipelago/pipeline.py [hunk 5]
```diff
@@ -0,0 +1,1028 @@
...
+    dirname = os.path.join(args.application, "bin")
+    netlist = os.path.join(dirname, "design.packed")
+    assert os.path.exists(netlist), netlist + " does not exist"
+    placement = os.path.join(dirname, "design.place")
+    assert os.path.exists(placement), placement + " does not exists"
+    route = os.path.join(dirname, "design.route")
+    assert os.path.exists(route), route + " does not exists"
+    return netlist, placement, route, args.frequency
+
+def load_netlist(netlist_file):
```
Comment (Kuree) [inline archipelago/pipeline.py:29]: How is this different from the function offered in `io.py`?

File: archipelago/pipeline.py [hunk 6]
```diff
@@ -0,0 +1,1028 @@
...
+        node1 = pnr_graph.Node("route", x, y, route_type = "REG", track = track, bit_width = bit_width, net_id = net_id, reg_name = reg_name)
+    elif segment[0] == "RMUX":
+        rmux_name, x, y, bit_width = segment[1:]
+        node1 = pnr_graph.Node("route", x, y, route_type = "RMUX", bit_width = bit_width, net_id = net_id, rmux_name = rmux_name)
+    else:
+        raise ValueError("Unrecognized route type")
+    return node1
+
+def get_tile_at(x, y, bw, placement, port = "", reg = False):
+    pond_ports = ["data_in_pond_0", "data_out_pond_0", "flush"]
```
Comment (Kuree) [inline archipelago/pipeline.py:113]: Ideally should be read out from the routing graph automatically.

File: archipelago/pipeline.py [hunk 7]
```diff
@@ -0,0 +1,1028 @@
...
+
+    for tile_id, place in placement.items():
```
Comment (Kuree) [inline archipelago/pipeline.py:115]: This doesn't look efficient. you have to search the entire placement for every x and y. Consider build a reverse-indexed placement first, e.g. (x, y) -> list[nodes]

File: archipelago/pipeline.py [hunk 8]
```diff
@@ -0,0 +1,1028 @@
...
+                if node2.route_type == "PORT":
+                    tile_id = get_tile_at(node2.x, node2.y, node2.bit_width, placement, node2.port)
+                    if tile_id[0] == "m" and not "chain" in node2.port:
+                         node2.reg = True
+                    graph.add_edge(node2, tile_id)
+                elif node2.route_type == "REG":
+                    tile_id = get_tile_at(node2.x, node2.y, node2.bit_width, placement, reg = True)
+                    node2.reg = True
+                    graph.add_edge(node2, tile_id)
+
```
Comment (Kuree) [inline archipelago/pipeline.py:182]: Too many white spaces. Please use pycodestyle to format the code.

File: archipelago/pipeline.py [hunk 9]
```diff
@@ -0,0 +1,1028 @@
...
+    
+    graph.update_sources_and_sinks()
+    graph.update_edge_kernels()
+    if 'PIPELINED' in os.environ and os.environ['PIPELINED'] == '1':
```
Comment (Kuree) [inline archipelago/pipeline.py:186]: Consider making it as a function argument and detect it in a centralized places.

File: archipelago/pipeline.py [hunk 10]
```diff
@@ -0,0 +1,1028 @@
...
+
+                tile_graph.edge_weights[(source, dest)] = num_avail_regs
+    
+    return tile_graph
+
+def verify_graph(graph):
+    while graph.FixCycles():
+        pass
+    
+PE_DELAY = 700
```
Comment (Kuree) [inline archipelago/pipeline.py:249]: Consider make it a simple configuration file instead of hard coded in

File: archipelago/pipeline.py [hunk 11]
```diff
@@ -0,0 +1,1028 @@
...
+            crit_edges.append((timing_info[curr_node].parent, curr_node))
+        curr_node = timing_info[curr_node].parent
+        if curr_node == None:
+            break
+        net_ids.add(graph.get_node(curr_node).net_id)
+        crit_path.append((curr_node, timing_info[curr_node].get_total()))
+
+    crit_path.reverse()
+
+    #print("\tCritical Path Nets:", *net_ids)
```
Comment (Kuree) [inline archipelago/pipeline.py:376]: Remove all commented out code

File: archipelago/pipeline.py [hunk 12]
```diff
@@ -0,0 +1,1028 @@
...
+
+def break_crit_path(graph, id_to_name, crit_path, placement, routes):
+    break_idx = find_break_idx(graph, crit_path)
+
+    break_node_source = crit_path[break_idx][0]
+    break_node_dest = graph.sinks[break_node_source][0]
+    g_break_node_source = graph.get_node(break_node_source)
+    g_break_node_dest = graph.get_node(break_node_dest)
+
+    assert g_break_node_source.type_ == "route"
```
Comment (Kuree) [inline archipelago/pipeline.py:419]: Ideally we should add a assertion message to explain why it's the case.

File: archipelago/pipeline.py [hunk 13]
```diff
@@ -0,0 +1,1028 @@
...
+
+def reg_into_route(routes, g_break_node_source, new_reg_route_source):
+    for net_id, net in routes.items():
+        for route in net:
+            for idx, segment in enumerate(route):
+                if g_break_node_source.to_route() == segment:
+                    route.insert(idx + 1, new_reg_route_source.to_route())
+                    return 
+
+def break_crit_path(graph, id_to_name, crit_path, placement, routes):
```
Comment (Kuree) [inline archipelago/pipeline.py:411]: Why are these functions not inside the graph class?

File: archipelago/pipeline.py [hunk 14]
```diff
@@ -0,0 +1,1028 @@
...
+
+    while True:
+        if graph.get_node(crit_path[break_idx + 1][0]).route_type == "RMUX" and graph.get_node(crit_path[break_idx][0]).route_type == "SB":
+            return break_idx
+        break_idx += 1
+
+        if break_idx + 1 >= len(crit_path):
+            break_idx = crit_path_adjusted.index(min(crit_path_adjusted))
+
+            while True:
```
Comment (Kuree) [inline archipelago/pipeline.py:395]: I'm not sure I'm following the logic. are these two level of while true loop necessary? can't you just use two separate for loops? Nested loop structures should be avoided if possible since it decrease the code readability. Same thing goes for other logic in the PR.

File: archipelago/pipeline.py [hunk 15]
```diff
@@ -0,0 +1,1028 @@
...
+            if parent not in node_cycles:
+                c = 0
+            else:
+                c = node_cycles[parent]
+            
+            if g_node.type_ == "route":
+                if graph.node_latencies[node] != 0:
+                    if len(graph.sinks[node]) > 0:
+                        child = graph.sinks[node][0]
+                        if (child not in id_to_name or "d_reg_" not in id_to_name[child]) and (parent not in graph.get_mems()):
```
Comment (Kuree) [inline archipelago/pipeline.py:546]: Not sure if it's a good way to test out names. The only thing you have control over is the ID name, which is maintained internally and not visible to normal users. That's why IDs are assigned based on their types. What if there is a name actually called `d_reg_`. Don't have to fix it this time but might need it fixed in the future.

File: archipelago/pipeline.py [hunk 16]
```diff
@@ -0,0 +1,1028 @@
...
+        src = route[0]
+        for seg_index, segment in enumerate(route):
+            fout.write(f"Segment: {seg_index} Size: {len(segment)}\n")
+
+            for node in segment:
+                fout.write(f"{segment_node_to_string(node)}\n")
+        fout.write("\n")
+
+
+def dump_placement_result(dir_name, placement, id_to_name):
```
Comment (Kuree) [inline archipelago/pipeline.py:946]: Those are in `io.py`

File: archipelago/pnr_graph.py [hunk 17]
```diff
@@ -0,0 +1,592 @@
...
+        node1 = edge[1]
+
+        if edge in self.edges:
+            self.edges.remove(edge)
+        if node0 in self.sources[node1]:
+            self.sources[node1].remove(node0)
+        if node1 in self.sinks[node0]:
+            self.sinks[node0].remove(node1)
+
+    def isCyclicUtil(self, v, visited, recStack):
```
Comment (Kuree) [inline archipelago/pnr_graph.py:378]: Make you naming consistent. Use snake case here.

File: archipelago/pnr_graph.py [hunk 18]
```diff
@@ -0,0 +1,592 @@
...
+class KernelNode:
+    def __init__(self, mem_id = None, kernel = None, type_ = None, latency = 0, flush_latency = 0, has_shift_regs = False):
+        self.mem_id = mem_id
+        self.kernel = kernel
+        self.type_ = type_
+        self.latency = latency
+        self.flush_latency = flush_latency
+        self.has_shift_regs = has_shift_regs
+
+    def to_string(self):
```
Comment (Kuree) [inline archipelago/pnr_graph.py:478]: use `__repr__`

File: archipelago/sta.py (hunk 2 above)
Comment (Kuree) [reply archipelago/sta.py:42]: If it's inside a module, you can invoke it with `python -m archipelago.sta`

File: archipelago/visualize.py [hunk 19]
```diff
@@ -0,0 +1,443 @@
...
+        if src_port not in self.next:
+            self.next[src_port] = []
+        self.next[src_port].append((node, sink_port))
+        node.parent[sink_port] = self
+
+    def __repr__(self):
+        return self.blk_id
+
+
+class Graph:
```
Comment (Kuree) [inline archipelago/visualize.py:92]: Can't you reuse the pnr graph defined above?

**Review by Kuree — CHANGES_REQUESTED** [review]

File: archipelago/pnr_.py (hunk 1 above)
Comment (Kuree) [reply archipelago/pnr_.py:130]: This is causing the error here: https://buildkite.com/stanford-aha/aha-flow/builds/5981#1c4bd7bb-c181-40bf-b72f-c488c98d9c61/501-556

Comment (jack-melchert) [reply archipelago/pnr_.py:130]: I've already changed the upstream so this should be fixed
